scripts/data_prep/validate_and_tokenize_data.py

```diff
@@ -0,0 +1,731 @@
+# MAGIC
+# MAGIC ### Note on Long-Term Solution:
+# MAGIC - **Temporary Measure**: This script is a stop-gap solution.
+# MAGIC - **Future Development**: We are in the process of developing a long-term data preparation service, which will eventually replace this script.
+# MAGIC
+# MAGIC ### Checks Include:
+# MAGIC - check input dataset:
+# MAGIC   1) verify if dataset input format is valid (need to be one of these: Huggingface, delta table, dbfs:/Volumes, cloud path);
+# MAGIC - check HF input location:
+# MAGIC   1) load dataset info and check if it is accessible;
+# MAGIC   2) verify if the split exists.
+# MAGIC - check cloud path location:
+# MAGIC   1) check the cloud prefix is compliant with composers' object store supports (gs, s3, oci)
+# MAGIC   2) check if list objects returns nothing.
+# MAGIC - count_tokens:
+# MAGIC   1) For IFT task: validate tokenization by running tokenizer + filter on the entire dataset. count the number of tokens. Throws error if there are any empty responses or prompts
+# MAGIC   2) For CPT task: call donwload_text_to_mds.py and count the resulted mds dataset. Note this could take a long time.
+# MAGIC
+# MAGIC ### Questions:
+# MAGIC - Is "download_text_to_mds.py" always callable from the validation script?
+# MAGIC - what is the function to reuse to run tokenization on HF datasets with filters?
+# MAGIC - The inputs to this validation script is assumed to be the same or a subset of the FT API arguments, i.e., a configuration like below. Is this a valid assumption?
+# MAGIC ```
+# MAGIC cfg = {
+# MAGIC     model: str,
+# MAGIC     train_data_path: str,
+# MAGIC     save_folder: str,
+# MAGIC     *,
+# MAGIC     task_type: Optional[str] = "INSTRUCTION_FINETUNE",
+# MAGIC     eval_data_path: Optional[str] = None,
+# MAGIC     eval_prompts: Optional[List[str]] = None,
+# MAGIC     custom_weights_path: Optional[str] = None,
+# MAGIC     training_duration: Optional[str] = None,
+# MAGIC     learning_rate: Optional[float] = None,
+# MAGIC     context_length: Optional[int] = None,
+# MAGIC     experiment_trackers: Optional[List[Dict]] = None,
+# MAGIC     data_prep_config: Optional[Dict] = None,
+# MAGIC     disable_credentials_check: Optional[bool] = None,
+# MAGIC     timeout: Optional[float] = 10,
+# MAGIC     future: Literal[False] = False,
+# MAGIC }
+# MAGIC - What null checkings do we want to have?
+# MAGIC - How to map the model to its expected eos_text / bos_text format? [Ref](https://databricks.slack.com/archives/C05K29T9NBF/p1703644153357929?thread_ts=1703643155.904289&cid=C05K29T9NBF)
+# MAGIC - How to automate tokenization for CPT? it is always really standard: sequence -> concat(tok(BOS), tok(sequence), tok(EOS)), and then concatenate sequences. [Ref](https://databricks.slack.com/archives/C05K29T9NBF/p1703698056000399?thread_ts=1703643155.904289&cid=C05K29T9NBF)
+# MAGIC ```
+
+# COMMAND ----------
+
+# MAGIC %pip install llm-foundry
```

Comment:
this should install the version of foundry that we're using in finetuning, so I think this will need to git clone a specific commit and pip install from the cloned repo